File: reference/goathal/backends/_refcompute.py
"""
Deterministic reference "compute" shared by the two reference backends.

This stands in for a real inference engine (llama.cpp / ONNX Runtime). The point of
the reference implementation is to exercise the PROTOCOL machinery (canonical
commitment, determinism profiles, cross-class comparison, conformance), not to
re-derive an inference stack. A real backend swaps these two functions for engine
calls behind the identical GoatBackend trait.

Key property for cross-class verification (Spec C):
  * tokens are IDENTICAL across backends (pinned greedy decode) -> token_agreement = 1.0
  * the numeric vector is computed from an identical base; a backend may add a small
    BOUNDED perturbation to model cross-vendor FP roundoff. l_inf(A,B) stays within the
    widened tolerance band.
Payload is treated as OPAQUE bytes: only hashed, never parsed for content.
"""
import hashlib

FP_SCALE = 1_000_000            # fixed-point vector range


def reference_tokens(payload: bytes, seed: int, n: int = 16):
    h = hashlib.sha3_256(b"tok" + payload + seed.to_bytes(8, "big")).digest()
    toks, ctr = [], 0
    while len(toks) < n:
        h = hashlib.sha3_256(h + ctr.to_bytes(4, "big")).digest()
        for i in range(0, len(h), 2):
            toks.append(int.from_bytes(h[i:i + 2], "big") % 32000)
            if len(toks) >= n:
                break
        ctr += 1
    return tuple(toks)


def reference_vector_base(payload: bytes, seed: int, n: int = 8):
    h = hashlib.sha3_256(b"vec" + payload + seed.to_bytes(8, "big")).digest()
    return tuple(int.from_bytes(h[i * 4:(i + 1) * 4], "big") % FP_SCALE for i in range(n))


def bounded_perturbation(base, magnitude: int):
    """Deterministic per-index delta in [-magnitude, +magnitude]; models FP roundoff."""
    if magnitude == 0:
        return tuple(base)
    out = []
    for i, v in enumerate(base):
        d = ((i * 7 + 3) % (2 * magnitude + 1)) - magnitude
        out.append(v + d)
    return tuple(out)

File: goatcoin-rs/crates/goat-backends/src/refcompute.rs
//! Deterministic reference "compute" shared by the reference backends (DEVICE layer).
//! Stands in for a real inference engine; a production backend swaps these for engine calls
//! behind the identical GoatBackend trait. Payload is treated as OPAQUE bytes (only hashed).

use sha3::{Digest, Sha3_256};

pub const FP_SCALE: u32 = 1_000_000;

fn sha(parts: &[&[u8]]) -> [u8; 32] {
    let mut h = Sha3_256::new();
    for p in parts {
        h.update(p);
    }
    let d = h.finalize();
    let mut o = [0u8; 32];
    o.copy_from_slice(&d);
    o
}

pub fn reference_tokens(payload: &[u8], seed: u64, n: usize) -> Vec<u32> {
    let mut out = Vec::with_capacity(n);
    let mut h = sha(&[b"tok", payload, &seed.to_be_bytes()]);
    let mut ctr: u32 = 0;
    while out.len() < n {
        h = sha(&[&h, &ctr.to_be_bytes()]);
        let mut i = 0;
        while i + 2 <= h.len() && out.len() < n {
            let v = u16::from_be_bytes([h[i], h[i + 1]]) as u32 % 32000;
            out.push(v);
            i += 2;
        }
        ctr += 1;
    }
    out
}

pub fn reference_vector_base(payload: &[u8], seed: u64, n: usize) -> Vec<i64> {
    let h = sha(&[b"vec", payload, &seed.to_be_bytes()]);
    (0..n)
        .map(|i| {
            let b = [h[i * 4], h[i * 4 + 1], h[i * 4 + 2], h[i * 4 + 3]];
            (u32::from_be_bytes(b) % FP_SCALE) as i64
        })
        .collect()
}

/// Deterministic per-index delta in [-magnitude, +magnitude]; models FP roundoff.
pub fn bounded_perturbation(base: &[i64], magnitude: i64) -> Vec<i64> {
    if magnitude == 0 {
        return base.to_vec();
    }
    base.iter()
        .enumerate()
        .map(|(i, &v)| {
            let d = ((i as i64 * 7 + 3) % (2 * magnitude + 1)) - magnitude;
            v + d
        })
        .collect()
}
